### ISSUE TITLE
Serialization and RPC libraries

### ISSUE THREAD

**Issue by tinybike** [ethereumjs/ideas#1]
With a little extra work, augur's [serialization](https://github.com/AugurProject/augur-abi) (horribly misnamed, as this is our implementation of the contract ABI, not "augur's ABI") and [RPC](https://github.com/AugurProject/ethrpc) libraries could possibly be useful general-purpose tools.

[augur-abi](https://github.com/AugurProject/augur-abi): needs Solidity support; right now only does contact ABI serialization for Serpent's datatypes.

[ethrpc](https://github.com/AugurProject/ethrpc): there are a few augur-specific things (such as the hosted nodes) that might need to be changed.

No idea if there's any interest in this, but just thought I'd throw this out there to see.

**Comment by kumavis** [comment]
`augur-abi` is a `serpent -> abi` generator? is there a `solidity -> abi` generator in js/emscripten currently?

**Comment by wanderer** [comment]
pluging @axic 's https://github.com/axic/ethereumjs-abi

**Comment by wanderer** [comment]
also think a minimalistic RPC wrapper would be great. Would it be able to work with different providers?

**Comment by tinybike** [comment]
@kumavis there's one buried in web3.  I copied it out for doing unit tests (comparing output): https://github.com/AugurProject/augur-abi/tree/master/lib/solidity.  It would be trivial to call to it in the actual program to do solidity serialization.  But we only use serpent at Augur so I've never gotten around to doing it :)

**Comment by axic** [comment]
@tinybike isn't the ABI the same at this stage apart from how it is described to the end user?

**Comment by axic** [comment]
@tinybike and as @wanderer mentioned, my -abi libary works with the notation described in the "[Contract ABI](https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI)"

**Comment by tinybike** [comment]
@axic Yes, the only difference that I'm aware of is the data types.  However if I remember correctly, Serpent's data types are a subset of Solidity's.  So if your -abi library works already, there's probably no need to involve augur-abi at all :)

**Comment by tinybike** [comment]
It looks like augur-abi is unnecessary, but ethrpc might be useful.  It does basic HTTP and IPC RPC (fully asynchronous) stuff.  It has two main additional features worth mentioning:

- ethrpc.transact does a send_transaction, followed by a call (to get return values), followed by polling for the transaction's blocknumber being set as "confirmation".

- It does some simple load balancing between augur's 4 hosted nodes.  Might be easiest to just move the load balancing stuff back to augur.js.

The main thing it is missing is a good wrapper for filters; our filter code is still part of augur.js.  It's modular though, so it should be straightforward to move it to ethrpc.

**Comment by axic** [comment]
Re: ethrpc, it would be great to have it as an independent lightweight tool. I was going to work on a simple RPC translator, but in my case (axic/tinyoracle), json-rpc2 worked well too.

**Comment by axic** [comment]
@tinybike I've just added Serpent data type support, would you be able to test it with Augur?

**Comment by tinybike** [comment]
Sure!  I'll run it through the test suite.

**Comment by tinybike** [comment]
IIRC bytes and string are the same type in serpent, denoted by 's'.  Will have to double check this though (am on the road at the moment)!

**Comment by axic** [comment]
@tinybike pushed it to here: https://github.com/axic/ethereumjs-abi/tree/feature/serpent

**Comment by kumavis** [comment]
im building `provider-engine`, a sort of skeleton for making web3 'zero-client' providers -- zero-client being a term i made up to refer to an ethereum node that just trustingly accepts blockchain data from some source without light client proofs. The main reason I made this is to get around some of the perceived issues of using the RPC to service many concurrent users (stateful filter-based methods, expensive evm run calls, etc).

`provider-engine` lets you compose a web3 provider from many 'sub-providers' that handle some subset of rpc methods, and can re-emit additional rpc's to facilitate the handling of the original request (e.g. `eth_call` -> `eth_getBalance`, `eth_getCode`, `eth_getStorage` ). `provider-engine` also handles caching of rpc requests at the top level. This is backed by an ethereum node behind a proxy that white-lists incomming rpc methods and caches results.

its designed to be very pluggable + modular, so you can just use what you need, or roll in some home-spun optimization.

its not ready yet but you can follow progress here
https://github.com/MetaMask/provider-engine

**Comment by kumavis** [comment]
@tinybike curious what augur's full stack looks like

**Comment by tinybike** [comment]
@kumavis Augur's full stack: http://docs.augur.net/#architecture

**Comment by stevenroose** [comment]
Any news on a serialization library for Solidity?

**Comment by axic** [comment]
@stevenroose I think this is the wrong channel.  If you are looking for an ABI encoder/decoder, check out https://github.com/etheremujs/ethereumjs-abi.

If you are looking for a library written in Solidity, please check out https://gitter.im/ethereum/solidity
